packages/arg-parser/src/arg-parser.ts

```diff
@@ -1,126 +1,142 @@
+import {
+  argMetadata,
+  getArgumentMetadata,
+  getDeprecatedArgsWithReplacement,
+  getUnsupportedArgs,
+  UnknownCliArgumentError,
+  UnsupportedCliArgumentError,
+} from './arg-metadata';
+
+function unwrapType(type: unknown): unknown {
+  if (type instanceof z.ZodOptional) {
+    return unwrapType(type.unwrap());
+  }
+  if (type instanceof z.ZodDefault) {
+    return unwrapType(type.unwrap());
+  }
+  return type;
+}
 
 /**
- * The yargs-parser options configuration.
+ * Generate yargs-parser configuration from schema
  */
-const OPTIONS = {
-  string: [
-    'apiVersion',
-    'authenticationDatabase',
-    'authenticationMechanism',
-    'awsAccessKeyId',
-    'awsIamSessionToken',
-    'awsSecretAccessKey',
-    'awsSessionToken',
-    'awsIamSessionToken',
-    'browser',
-    'csfleLibraryPath',
-    'cryptSharedLibPath',
-    'db',
-    'gssapiHostName',
-    'gssapiServiceName',
-    'sspiHostnameCanonicalization',
-    'sspiRealmOverride',
-    'jsContext',
-    'host',
-    'keyVaultNamespace',
-    'kmsURL',
-    'locale',
-    'oidcFlows',
-    'oidcRedirectUri',
-    'password',
-    'port',
-    'sslPEMKeyFile',
-    'sslPEMKeyPassword',
-    'sslCAFile',
-    'sslCertificateSelector',
-    'sslCRLFile',
-    'sslDisabledProtocols',
-    'tlsCAFile',
-    'tlsCertificateKeyFile',
-    'tlsCertificateKeyFilePassword',
-    'tlsCertificateSelector',
-    'tlsCRLFile',
-    'tlsDisabledProtocols',
-    'username',
-  ],
-  boolean: [
-    'apiDeprecationErrors',
-    'apiStrict',
-    'buildInfo',
-    'deepInspect',
-    'exposeAsyncRewriter',
-    'help',
-    'ipv6',
-    'nodb',
-    'norc',
-    'oidcTrustedEndpoint',
-    'oidcIdTokenAsAccessToken',
-    'oidcNoNonce',
-    'perfTests',
-    'quiet',
-    'retryWrites',
-    'shell',
-    'smokeTests',
-    'skipStartupWarnings',
-    'ssl',
-    'sslAllowInvalidCertificates',
-    'sslAllowInvalidHostnames',
-    'sslFIPSMode',
-    'tls',
-    'tlsAllowInvalidCertificates',
-    'tlsAllowInvalidHostnames',
-    'tlsFIPSMode',
-    'tlsUseSystemCA',
-    'verbose',
-    'version',
-  ],
-  array: ['eval', 'file'],
-  alias: {
-    h: 'help',
-    p: 'password',
-    u: 'username',
-    f: 'file',
-    'build-info': 'buildInfo',
-    json: 'json', // List explicitly here since it can be a boolean or a string
-    browser: 'browser', // ditto
-    oidcDumpTokens: 'oidcDumpTokens', // ditto
-    oidcRedirectUrl: 'oidcRedirectUri', // I'd get this wrong about 50% of the time
-    oidcIDTokenAsAccessToken: 'oidcIdTokenAsAccessToken', // ditto
-  },
-  configuration: {
+export function generateYargsOptionsFromSchema({
+  schema,
+  configuration = {
     'camel-case-expansion': false,
     'unknown-options-as-args': true,
     'parse-positional-numbers': false,
     'parse-numbers': false,
     'greedy-arrays': false,
     'short-option-groups': false,
   },
-};
+}: {
+  schema: z.ZodObject;
+  configuration?: YargsOptions['configuration'];
+}): YargsOptions {
```

Comment:
It sucks that we can't easily do `unwrapType()` typesafely (right?), but the fact that we have to rely on casting to `unknown` and on `instanceof` here means we have very little guarantee that an incorrect input type would be properly handled.

Can we make an assertion here that `schema instanceof z.ZodObject`? Otherwise this would just silently fail if we get an object created from a different zod library, which I don't think we'd want.